src/query/ts/m3db/encoded_unconsolidated_step_iterator.go

```diff
@@ -21,99 +21,150 @@
-	block.UnconsolidatedStep,
+func (b *encodedBlockUnconsolidated) StepIter() (
+	block.UnconsolidatedStepIter,
 	error,
 ) {
-	it.mu.RLock()
-	if it.exhausted {
-		it.mu.RUnlock()
-		return nil, fmt.Errorf("out of bounds")
-	}
+	cs := b.consolidation
+	accumulator := consolidators.NewStepLookbackAccumulator(
+		b.lookback,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+	)
+
+	return &encodedStepIterUnconsolidated{
+		idx:            -1,
+		meta:           b.meta,
+		seriesMeta:     b.seriesMetas,
+		seriesIters:    b.seriesBlockIterators,
+		lastBlock:      b.lastBlock,
+		expandedSeries: make([][]xts.Datapoints, len(b.seriesBlockIterators)),
+
+		lookback:    b.lookback,
+		accumulator: accumulator,
+		seriesPeek:  make([]peekValue, len(b.seriesBlockIterators)),
+	}, nil
+}
 
-	if it.err != nil {
-		it.mu.RUnlock()
-		return nil, it.err
+func (it *encodedStepIterUnconsolidated) Current() block.UnconsolidatedStep {
+	points := it.accumulator.AccumulateAndMoveToNext()
+	return storage.NewUnconsolidatedStep(it.stepTime, points)
+}
+
+func (it *encodedStepIterUnconsolidated) nextForStep(
+	i int,
+	stepTime time.Time,
+) {
+	peek := it.seriesPeek[i]
+	if peek.finished {
+		// No next value in this iterator
+		return
 	}
 
-	var t time.Time
-	values := make([]xts.Datapoints, len(it.seriesIters))
-	for i, iter := range it.seriesIters {
-		// Skip this iterator if it's not valid
-		if !it.validIters[i] {
-			continue
+	if peek.started {
+		point := peek.point
+		if point.Timestamp.After(stepTime) {
+			// This point exists further than the current step
+			// There are next values, but current step is empty.
+			return
 		}
 
-		dp, _, _ := iter.Current()
-		if i == 0 {
-			t = dp.Timestamp
-		}
+		// clear peeked point.
+		it.seriesPeek[i].started = false
+		// Currently at a potentially viable data point.
+		// Record previously peeked value, and all potentially valid values.
+		it.accumulator.AddPointForIterator(point, i)
 
-		values[i] = []xts.Datapoint{{
-			Timestamp: dp.Timestamp,
-			Value:     dp.Value,
-		}}
+		// If at boundary, add the point as the current value.
+		if point.Timestamp.Equal(stepTime) {
+			return
+		}
 	}
 
-	step := storage.NewUnconsolidatedStep(t, values)
-	it.mu.RUnlock()
-	return step, nil
+	iter := it.seriesIters[i]
+	// Read through iterator until finding a data point outside of the
+	// range of this step; then set the next peek value.
+	for iter.Next() {
+		dp, _, _ := iter.Current()
+
+		// If this datapoint is before the current timestamp, add it to
+		// the accumulator
+		if !dp.Timestamp.After(stepTime) {
+			it.seriesPeek[i].started = false
+			it.accumulator.AddPointForIterator(dp, i)
+		} else {
+			// This point exists further than the current step; set peeked value
+			// to this point.
+			it.seriesPeek[i].point = dp
+			it.seriesPeek[i].started = true
+			return
+		}
+	}
 }
 
 func (it *encodedStepIterUnconsolidated) Next() bool {
```

Comment:
Given that the `Next()` functions are pretty similar for series and step, think it's worth moving the common parts into a common function?